torchao/csrc/metal/test_int4_q_kernels.mm

```diff
@@ -0,0 +1,291 @@
+    lib_name = lib_name_;
+    lib = compileLibraryFromFile(device, lib_name + ".metal");
+  }
+  Int4MMBase(id<MTLDevice> device, unsigned M_, unsigned N_, unsigned K_)
+      : M(M_), N(N_), K(K_), lib(nil) {
+    allocBuffers(device);
+  }
+
+  virtual void dispatchThreads(id<MTLComputeCommandEncoder> encoder,
+                               unsigned maxThreadsPerGroup) const {}
+
+  void encodeMM(id<MTLCommandBuffer> cmdBuffer,
+                id<MTLComputePipelineState> cpl) const {
+    id<MTLComputeCommandEncoder> encoder = [cmdBuffer computeCommandEncoder];
+    std::vector<unsigned> sizes = {M, K, N, 0};
+    const auto maxThreadsPerGroup =
+        static_cast<decltype(M)>([cpl maxTotalThreadsPerThreadgroup]);
+    [encoder setComputePipelineState:cpl];
+    [encoder setBuffer:buf_A offset:0 atIndex:0];
+    [encoder setBuffer:buf_B offset:0 atIndex:1];
+    [encoder setBuffer:buf_SZ offset:0 atIndex:2];
+    [encoder setBuffer:buf_C offset:0 atIndex:3];
+    [encoder setBytes:sizes.data()
+               length:sizeof(uint32_t) * sizes.size()
+              atIndex:4];
+    dispatchThreads(encoder, maxThreadsPerGroup);
+    [encoder endEncoding];
+  }
+
+  template <typename T> void init() {
+    T *a_ptr = reinterpret_cast<T *>([buf_A contents]);
+    uint8_t *b_ptr = reinterpret_cast<uint8_t *>([buf_B contents]);
+    T *c_ptr = reinterpret_cast<T *>([buf_C contents]);
+    T *s_ptr = reinterpret_cast<T *>([buf_SZ contents]);
+    std::random_device rd;
+    std::mt19937 generator(rd());
+    std::uniform_int_distribution<> int_distrib(-8, 7);
+    std::uniform_real_distribution<> real_distrib(-1.0, 1.0);
+
+    for (unsigned idx = 0; idx < M * K; ++idx) {
+      a_ptr[idx] = real_distrib(generator);
+    }
+    for (unsigned idx = 0; idx < N * K / 2; ++idx) {
+      int32_t b0 = int_distrib(generator);
+      int32_t b1 = int_distrib(generator);
+      b_ptr[idx] = ((b1 + 8) << 4) | (b0 + 8);
+    }
+    for (unsigned idx = 0; idx < N * K / groupSize; ++idx) {
+      s_ptr[2 * idx] = (idx + 1.0) / N;
+      s_ptr[2 * idx + 1] = 0;
+    }
+    for (unsigned idx = 0; idx < M * N; ++idx) {
+      c_ptr[idx] = -1.0;
+    }
+  }
+
+  template <typename T>
+  bool validate(float atol_lim = 5e-4, float rtol_lim = 5e-3) const {
+    T *a_ptr = reinterpret_cast<T *>([buf_A contents]);
+    uint8_t *b_ptr = reinterpret_cast<uint8_t *>([buf_B contents]);
+    T *c_ptr = reinterpret_cast<T *>([buf_C contents]);
+    T *sz_ptr = reinterpret_cast<T *>([buf_SZ contents]);
+
+    for (unsigned m = 0; m < M; m++) {
+      for (unsigned n = 0; n < N; n++) {
+        float expected = float(c_ptr[m * N + n]);
+        const uint32_t k_block = (K + groupSize - 1) / groupSize;
+        const T *A_ptr = a_ptr + m * K;
+
+        float rc = 0.0;
+        uint k = 0;
+        for (uint32_t kb = 0; kb < k_block; kb++) {
+          const T scale = sz_ptr[(kb * N + n) * 2 + 0];
+          const T zero = sz_ptr[(kb * N + n) * 2 + 1] - scale * T(8);
```

Comment:
array indexing has always been tricky for me, but doesn't this imply that the layout of scalesAndZeros is [groupsize, N, 2] rather than [N, groupsize, 2]? we are skipping N elements per group, indicating that there are N columns, right?

Reply:
> doesn't this imply that the layout of scalesAndZeros

you are exactly right. My comment was wrong. i have fixed that and updating the pR now.